File: perl6/021.pl
#!/usr/bin/perl6

my %map;
for 0 .. 9999 {
  %map{$_} = sum_factors($_);
}

my $total;
for %map.kv -> $key, $value 
{
  $total += $key if ((defined %map{$value}) && ($key != $value) && (%map{$value} == $key));
}
say $total;

sub sum_factors($number) {
  my $sum = 1;
  loop (my $i = 2; $i <= sqrt($number); $i++) {
    if (!($number % $i)) {
      $sum += ($i + ($number / $i));
    }
  }
  return $sum;
}
  

File: javascript/021.js
"use strict"

const SIZE = 10000;

function sum_factors(number)
{
  let sum = 1;
  for (let i = 2; i <= Math.sqrt(number); ++i)
    if (!(number % i))
      sum += (i + (number / i));

  return sum;
}

let sumproperdivisors = new Array(SIZE);
for (let i=0; i<SIZE ; i++)
  sumproperdivisors[i] = sum_factors(i);

let total = 0;
for (let i=0; i<SIZE ; i++)
  if ((sumproperdivisors[i] < SIZE) && (i != sumproperdivisors[i]) && (sumproperdivisors[sumproperdivisors[i]] == i))
    total += i;

console.log(total);
